pkg/usagetracker/tracker_store.go

```diff
@@ -243,13 +266,42 @@ func (t *trackerStore) cleanup(now time.Time) {
 func (t *trackerStore) updateLimits() {
 	t.mtx.RLock()
 	tenantsClone := maps.Clone(t.tenants)
+	sortedTenants := slices.Clone(t.sortedTenants)
 	t.mtx.RUnlock()
 
 	zonesCount := t.limiter.zonesCount()
-	for tenantID, tenant := range tenantsClone {
+	var closeToLimit []string
+
+	for _, tenantID := range sortedTenants {
+		tenant := tenantsClone[tenantID]
 		limit := zeroAsNoLimit(t.limiter.localSeriesLimit(tenantID))
-		tenant.currentLimit.Store(currentSeriesLimit(tenant.series.Load(), limit, zonesCount))
+		series := tenant.series.Load()
+		tenant.currentLimit.Store(currentSeriesLimit(series, limit, zonesCount))
+
+		// Determine if this user is close to their limit.
+		// A user is close if: series >= (limit * percentageThreshold / 100)
+		if limit != noLimit {
+			percentageThreshold := limit * uint64(t.userCloseToLimitPercentageThreshold) / 100
+
+			if series >= percentageThreshold {
+				closeToLimit = append(closeToLimit, tenantID)
+			}
+		}
 	}
+
+	// Replace the users close to limit list atomically.
+	// We build a new slice every time instead of mutating the existing one.
+	t.mtx.Lock()
+	t.usersCloseToLimit = closeToLimit
```

Comment:
This isn't an atomic write, as the doc above says. It seems to me it's only safe to read this under `t.mtx`.